packages/diagrams/backend/sirius-components-collaborative-diagrams/src/main/java/org/eclipse/sirius/components/collaborative/diagrams/handlers/InvokeActionEventHandler.java

```diff
@@ -0,0 +1,190 @@
+
+    private final IObjectSearchService objectSearchService;
+
+    private final IRepresentationDescriptionSearchService representationDescriptionSearchService;
+
+    private final IDiagramDescriptionService diagramDescriptionService;
+
+    private final IDiagramQueryService diagramQueryService;
+
+    private final ICollaborativeDiagramMessageService messageService;
+
+    private final Counter counter;
+
+    public InvokeActionEventHandler(IObjectSearchService objectSearchService, IRepresentationDescriptionSearchService representationDescriptionSearchService, IDiagramDescriptionService diagramDescriptionService, IDiagramQueryService diagramQueryService,
+            ICollaborativeDiagramMessageService messageService, MeterRegistry meterRegistry) {
+        this.objectSearchService = Objects.requireNonNull(objectSearchService);
+        this.representationDescriptionSearchService = Objects.requireNonNull(representationDescriptionSearchService);
+        this.diagramDescriptionService = Objects.requireNonNull(diagramDescriptionService);
+        this.diagramQueryService = Objects.requireNonNull(diagramQueryService);
+        this.messageService = Objects.requireNonNull(messageService);
+
+        this.counter = Counter.builder(Monitoring.EVENT_HANDLER)
+                .tag(Monitoring.NAME, this.getClass().getSimpleName())
+                .register(meterRegistry);
+    }
+
+    @Override
+    public boolean canHandle(IDiagramInput diagramInput) {
+        return diagramInput instanceof InvokeActionInput;
+    }
+
+    @Override
+    public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IDiagramContext diagramContext, IDiagramInput diagramInput) {
+        this.counter.increment();
+
+        String message = this.messageService.invalidInput(diagramInput.getClass().getSimpleName(), InvokeActionInput.class.getSimpleName());
+        IPayload payload = new ErrorPayload(diagramInput.id(), message);
+        ChangeDescription changeDescription = new ChangeDescription(ChangeKind.NOTHING, diagramInput.representationId(), diagramInput);
+
+        if (diagramInput instanceof InvokeActionInput input) {
+            var diagram = diagramContext.getDiagram();
+            var actionId = input.actionId();
+            var optionalDiagramDescription = this.representationDescriptionSearchService.findById(editingContext, diagram.getDescriptionId())
+                    .filter(DiagramDescription.class::isInstance)
+                    .map(DiagramDescription.class::cast);
+            var optionalNode = this.diagramQueryService.findNodeById(diagram, input.diagramElementId());
+            var optionalNodeDescription = this.findNodeDescription(diagram, input.targetElementId(), optionalDiagramDescription.get(), optionalNode);
+            var optionalAction = this.getAction(actionId, optionalNodeDescription, optionalNode, editingContext, diagramContext, input.diagramElementId());
```

Comment:
Just find the diagramElement and use a service like this:

```
public interface IActionHandler {
  boolean canHandle(IEditingContext editingContext, IDiagramContext diagramContext, IDiagramElement diagramElement, actionId);
  Status handle(IEditingContext editingContext, IDiagramContext diagramContext, IDiagramElement diagramElement, actionId);
}
```

with a field in this class `private final List<IActionHandler> actionHandlers`.

It has several advantages over the current approach:

- One could write an action handler as a simple Spring service which would work for all diagrams
- One could overwrite the default behavior using Spring annotations to do something else

Reply:
Ok.
With this approach, how will it be possible to be sure to a specific IActionHandler will be taken first instead of the default IActionHandler that will work on all diagrams.
Please let me know if I need to add an IDefaultActionHandler to handle this "problem".

Reply:
If both answers yes, the problem is still there.
But I will do what you want.